### PR TITLE
#143 Handle Consul errors in a consistent fashion

### PR DESCRIPTION
**Pull request by chrsblck**
This PR changes the flow of the scraper to not bail out on specific API failures, after `consul` is deemed to be `up`. If one of the `consul` APIs fail, this metric datapoint will be null, which should work well with things like dashboards and/or monitoring (up to the user to treat missing as breaching or not).

### CODE REVIEW

File: consul_exporter.go [hunk 1]
```diff
@@ -213,21 +213,22 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
...
-	}
-	if len(leader) == 0 {
-		ch <- prometheus.MustNewConstMetric(
-			clusterLeader, prometheus.GaugeValue, 0,
-		)
 	} else {
-		ch <- prometheus.MustNewConstMetric(
-			clusterLeader, prometheus.GaugeValue, 1,
-		)
+		if len(leader) == 0 {
```
Comment (brian-brazil) [inline consul_exporter.go]: You can't use the return value if err != nill.

**Comment by chrsblck** [comment]
@simonpasquier I use Intellij + Go modules enabled, which keeps adding `.../go-modern/{concurrent,reflect2}`. I'll remove these in a bit, after I see why `test` and Intellij differ here.

**Comment by chrsblck** [comment]
> If there's any failures, we should be reporting consul as down. Partial data is not practical to deal with as a user.

Thanks for the review @brian-brazil. 

Do you think it the design change is wrong as well, or just make sure `consul_up=0` is reported in any of the error cases?

File: consul_exporter.go (hunk 1 above)
Comment (chrsblck) [reply consul_exporter.go]: Ah, good point. I'll update the specific `err` variable names so it's not shadowed throughout this function.

**Comment by chrsblck** [comment]
@simonpasquier I think this is ready for a review, thanks.

File: consul_exporter.go [hunk 2]
```diff
@@ -223,48 +239,56 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
...
-		return
+		log.Errorf("Failed to query for services: %v", err)
+		return false
 	}
 	ch <- prometheus.MustNewConstMetric(
 		serviceCount, prometheus.GaugeValue, float64(len(serviceNames)),
 	)
-
 	if e.healthSummary {
 		e.collectHealthSummary(ch, serviceNames)
```
Comment (simonpasquier) [inline consul_exporter.go]: This should also return whether the Consul request succeeded or not.

File: consul_exporter.go [hunk 3]
```diff
@@ -192,27 +191,43 @@ func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
...
 		)
-		log.Errorf("Can't query consul: %v", err)
-		return
 	}
 
-	// We'll use peers to decide that we're up.
-	ch <- prometheus.MustNewConstMetric(
-		up, prometheus.GaugeValue, 1,
-	)
+	e.collectKeyValues(ch)
```
Comment (simonpasquier) [inline consul_exporter.go]: could you move it with the other collect functions?

**Comment by chrsblck** [comment]
> you could do it a bit differently and have collectOneHealthSummary() return a bool like the other collect functions.

@simonpasquier Agreed, this is cleaner. I rebased this change into the last commit.

**Comment by simonpasquier** [comment]
Thanks!

**Pull request merged**